Support time travel in Hudi connector

albericgenius opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description
Fix #15003 
-  This PR only implement the incremental query for COW table.
-  Will discuss and implement the MOR table after MOR reading merged.
-  This PR depend on [the TPCH modification](https://github.com/trinodb/tpch/pull/32), and the PR have not passed yet. so the build probably will be failed. :)


<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues
https://issues.apache.org/jira/browse/HUDI-2692
https://issues.apache.org/jira/browse/HUDI-2693


<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`15003 `)
```

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java:
```diff
@@ -47,6 +48,8 @@
     private int maxSplitsPerSecond = Integer.MAX_VALUE;
     private int maxOutstandingSplits = 1000;
 
+    private String userInputInstantTimeFormat = DEFAULT_INSTANT_TIME_FORMAT;
```
nit: Remove a redundant space at L50.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java:
```diff
@@ -191,4 +194,17 @@ public HudiConfig setMaxOutstandingSplits(int maxOutstandingSplits)
+        return userInputInstantTimeFormat;
+    }
+
+    @Config("hudi.user-input-instant-time-format")
```
I would rename to `hudi.commit-time-format` or something.

albericgenius replied:
removed

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java:
```diff
@@ -191,4 +194,17 @@ public HudiConfig setMaxOutstandingSplits(int maxOutstandingSplits)
+    }
+
+    @Config("hudi.user-input-instant-time-format")
+    @ConfigDescription("Supported user input instantTime format for the incremental query.")
```
I'm wondering why Hudi allows a user-defined timestamp format in the metadata column. Can you share the official documentation or link to the source code?

albericgenius replied:
removed this config

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java:
```diff
@@ -191,4 +194,17 @@ public HudiConfig setMaxOutstandingSplits(int maxOutstandingSplits)
+    public void setUserInputInstantTimeFormat(String userInputInstantTimeFormat)
+    {
+        this.userInputInstantTimeFormat = userInputInstantTimeFormat;
+    }
```
Please update  `hudi.rst` and `TestHudiConfig`

albericgenius replied:
removed

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiErrorCode.java:
```diff
@@ -28,7 +28,9 @@
     HUDI_UNSUPPORTED_FILE_FORMAT(5, EXTERNAL),
-    HUDI_CURSOR_ERROR(6, EXTERNAL);
+    HUDI_CURSOR_ERROR(6, EXTERNAL),
+
```
Remove a redundant empty line.

albericgenius replied:
fixed

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiErrorCode.java:
```diff
@@ -28,7 +28,9 @@
-    HUDI_CURSOR_ERROR(6, EXTERNAL);
+    HUDI_CURSOR_ERROR(6, EXTERNAL),
+
+    HUDI_INVALID_UER_INPUT_INSTANT_TIME_FORMAT(7, EXTERNAL);
```
I expect "the config value is wrong" from this naming, but this is used for handling failure when parsing input value from Hudi.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSessionProperties.java:
```diff
@@ -48,6 +51,8 @@
     private static final String STANDARD_SPLIT_WEIGHT_SIZE = "standard_split_weight_size";
     private static final String MINIMUM_ASSIGNED_SPLIT_WEIGHT = "minimum_assigned_split_weight";
 
+    private static final String USER_INPUT_INSTANT_TIME_FORMAT = "user_input_instant_time_format";
```
nit: Remove a redundant empty line at L53.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSessionProperties.java:
```diff
@@ -116,7 +126,7 @@ public List<PropertyMetadata<?>> getSessionProperties()
     public static List<String> getColumnsToHide(ConnectorSession session)
     {
-        return (List<String>) session.getProperty(COLUMNS_TO_HIDE, List.class);
+        return (List<String>) session.getProperty(COLUMNS_TO_HIDE, List.class).stream().filter(column -> !INSTANT_TIME.equals(column)).collect(toUnmodifiableList());
```
I don't think hard-coded `INSTANT_TIME` filter is a good idea. Please revert this change.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSplitManager.java:
```diff
@@ -84,6 +84,7 @@ public ConnectorSplitSource getSplits(
             Constraint constraint)
     {
         HudiTableHandle hudiTableHandle = (HudiTableHandle) tableHandle;
+
```
Revert unrelated change.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiUtil.java:
```diff
@@ -25,29 +26,42 @@
 
 public final class HudiUtil
 {
+    public static final String INSTANT_TIME = "_hoodie_commit_time";
```
Rename to `HUDI_COMMIT_TIME_COLUMN_NAME`.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiUtil.java:
```diff
@@ -25,29 +26,42 @@
 public final class HudiUtil
 {
+    public static final String INSTANT_TIME = "_hoodie_commit_time";
+    public static final String TPCH_INSTANT_TIME = "n__hoodie_commit_time";
```
Test specific constant shouldn't exist here.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiUtil.java:
```diff
@@ -161,4 +175,47 @@ public static FileStatus getFileStatus(HoodieBaseFile baseFile)
+    {
+        try {
+            if (!DEFAULT_INSTANT_TIME_FORMAT.equals(userInputFormat)) {
+                SimpleDateFormat formatter = new SimpleDateFormat(userInputFormat);
```
It would be better to verify whether the format is valid in `HudiConfig`.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiUtil.java:
```diff
@@ -161,4 +175,47 @@ public static FileStatus getFileStatus(HoodieBaseFile baseFile)
+                formatter.applyPattern(DEFAULT_INSTANT_TIME_FORMAT);
+                return formatter.format(userInputTime);
+            }
+            else {
```
Remove a redundant else.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiUtil.java:
```diff
@@ -161,4 +175,47 @@ public static FileStatus getFileStatus(HoodieBaseFile baseFile)
+                return timeStamp;
+            }
+        }
+        catch (ParseException e) {
```
`ParseException` won't be thrown then the condition goes to `else`. Please minimize the try-catch block. 
Also,

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiUtil.java:
```diff
@@ -161,4 +175,47 @@ public static FileStatus getFileStatus(HoodieBaseFile baseFile)
+        boolean predicate = true;
+        for (Range range : ranges) {
+            if (range.getLowValue().isPresent()) {
+                predicate = HoodieTimeline.compareTimestamps(
```
nit: Static import `compareTimestamps`

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiUtil.java:
```diff
@@ -161,4 +175,47 @@ public static FileStatus getFileStatus(HoodieBaseFile baseFile)
+                        GREATER_THAN,
+                        convertToHoodieInstantTimeFormat(
+                                userInstantTimeFormat,
+                                new String(((Slice) range.getLowBoundedValue()).getBytes(), StandardCharsets.UTF_8)));
```
nit: Static import `UTF_8`.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java:
```diff
@@ -55,7 +61,10 @@
 
     private List<String> hivePartitionNames;
 
+    private List<String> commitTimeRanges;
```
Add `final`

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java:
```diff
@@ -55,7 +61,10 @@
+    private List<String> commitTimeRanges;
+
     public HudiReadOptimizedDirectoryLister(
+            ConnectorSession session,
```
We don't need the entire `ConnectorSession`. Only timestamp format should be sufficient.

albericgenius replied:
fixed

ebyhr commented on plugin/trino-hudi/src/test/java/io/trino/plugin/hudi/TestHudiCopyOnWriteConnectorTest.java:
```diff
@@ -32,4 +36,25 @@ protected QueryRunner createQueryRunner()
+    @Test(dataProvider = "testIncrementalQuery")
+    public void testCopyOnWriteIncrementalQuery(@Language("SQL") String query)
+    {
+        assertQuery(query);
```
Could you add product tests with non-default timestamp format?

ebyhr commented on testing/trino-server-dev/etc/catalog/hudi.properties:
```diff
@@ -12,4 +12,5 @@
 # limitations under the License.
 #
 connector.name=hudi
+hudi.columns-to-hide=_hoodie_commit_time,_hoodie_commit_seqno,_hoodie_record_key,_hoodie_partition_path,_hoodie_file_name
```
Please revert.

albericgenius replied:
reverted

ebyhr commented on testing/trino-testing/src/main/java/io/trino/testing/AbstractTestQueryFramework.java:
```diff
@@ -99,13 +99,18 @@ public void init()
     {
         queryRunner = afterClassCloser.register(createQueryRunner());
-        h2QueryRunner = afterClassCloser.register(new H2QueryRunner());
+        h2QueryRunner = afterClassCloser.register(createBaseH2QueryRunner());
```
Revert changes in this class.

albericgenius replied:
reverted

ebyhr commented on testing/trino-testing/src/main/java/io/trino/testing/H2QueryRunner.java:
```diff
@@ -91,13 +91,19 @@
         implements Closeable
 {
-    private final Handle handle;
+    protected final Handle handle;
```
Revert changes in this class.

albericgenius replied:
reverted

ebyhr commented:
> Fix #15003

This isn't a time travel feature. Please take a look at Iceberg page in the issue.

electrum commented on plugin/trino-hudi/pom.xml:
```diff
@@ -429,6 +429,12 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
```
Why do we need to add this here?

albericgenius replied:
reverted

codope commented:
> > Fix #15003
> 
> This isn't a time travel feature. Please take a look at Iceberg page in the issue.

@ebyhr Thanks for pointing to the commit. Time travel is just a fancy name for a versioned snapshot query. While we're using `_hoodie_commit_time` as a filter in this PR, `AS OF` semantics in SQL is much better suited for this feature. I couldn't find `SELECT ... AS OF` in Trino [SQL docs](https://trino.io/docs/current/sql/select.html). Is it only supported for Iceberg connector?

ebyhr commented:
>  Is it only supported for Iceberg connector?

@codope Yes.

codope commented:
Got it. I think this commit is more relevant for us https://github.com/trinodb/trino/commit/1c23c6976636768db254bd77807b3512fda96dbb which adds the necessary connector metadata APIs to support versioned snapshot queries. If we implement those APIs in `HudiMetadata`, we can run `SELECT ... AS OF` queries through Hudi connector.

ebyhr commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/HudiPushTableVersionIntoTableScan.java:
```diff
@@ -0,0 +1,131 @@
+import static io.trino.sql.planner.plan.Patterns.tableScan;
+import static java.util.Objects.requireNonNull;
+
+public class HudiPushTableVersionIntoTableScan
```
Hudi specific code shouldn't exist in the planner.

albericgenius replied:
Removed the Hudi specific code in new revision.

albericgenius commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/HudiPushTableVersionIntoTableScan.java:
```diff
@@ -0,0 +1,131 @@
+        return Result.empty();
+    }
+
+    public static TupleDomain<ColumnHandle> computeEnforced(TupleDomain<ColumnHandle> predicate, TupleDomain<ColumnHandle> unenforced)
```
I will removed this unused static function

albericgenius replied:
Fixed

maddy2u commented:
When are we expecting this to be merged ?

ebyhr commented:
It depends on @albericgenius. I've waited that @albericgenius addresses comments.

albericgenius commented:
> It depends on @albericgenius. I've waited that @albericgenius addresses comments.

Got you, I will continue to work on this.

maddy2u commented:
Hey @albericgenius  - Any update on this?

albericgenius commented:
> Hey @albericgenius - Any update on this?

I updated a new revision, removed the Hudi specific code in the planner

@ebyhr @electrum @codope Thanks for your coaching and time.

ebyhr commented on plugin/trino-hudi/src/test/java/io/trino/plugin/hudi/TestHudiSmokeTest.java:
```diff
@@ -165,6 +167,31 @@ public void testPartitionColumn()
+    public void testTimeTravelQuery(@Language("SQL") String sql, String expected)
+    {
+        assertQuery(sql, expected);
+    }
```
Could you add a test case for schema evolution in Hudi (e.g. add / drop column)?

albericgenius replied:
Hudi connector have not supported add/ drop column yet.
I am working on the schema revolution for hudi query now(plan to submit PR 2 or 3 weeks later).
Is possible to add a test case for schema evolution in the schema revolution PR?

ebyhr replied:
I mean schema evolution on Spark Hudi.

albericgenius replied:
got you, i will update

albericgenius replied:
added one schema evolution case on Spark Hudi.

ebyhr reviewed (commented):
Could you add "Time travel queries" section to `hudi.rst` likes Iceberg documentation? It would be nice to mention the way to find versions from existing Hudi tables.

albericgenius commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiTableHandle.java:
```diff
@@ -40,13 +50,21 @@ public HudiTableHandle(
         this.basePath = requireNonNull(basePath, "basePath is null");
         this.tableType = requireNonNull(tableType, "tableType is null");
+        this.partitionColumns = requireNonNull(partitionColumns, "partitionColumns is null");
+        this.dataColumns = requireNonNull(dataColumns, "dataColumns is null");
```
Fixed and updated the hudi.rst

albericgenius commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiUtil.java:
```diff
@@ -161,4 +176,45 @@ public static FileStatus getFileStatus(HoodieBaseFile baseFile)
         }
     }
+
+    public static HoodieTableMetaClient buildTableMetaClient(Configuration configuration, String basePath)
```
thanks for detail info, fixed

albericgenius commented on plugin/trino-hudi/src/test/java/io/trino/plugin/hudi/TestHudiSmokeTest.java:
```diff
@@ -165,6 +167,31 @@ public void testPartitionColumn()
     }
 
+    @DataProvider(name = "timeTravelQueryProvider")
+    public Object[][] timeTravelQueryProvider()
```
fixed

albericgenius commented on plugin/trino-hudi/src/test/java/io/trino/plugin/hudi/BaseHudiConnectorTest.java:
```diff
@@ -95,4 +97,23 @@ protected static String columnsToHide()
     }
+
+    @Test
+    @Override
```
Thanks for your time, I didn't notice that function before. :)

ebyhr commented on docs/src/main/sphinx/connector/hudi.rst:
```diff
@@ -157,3 +157,25 @@ Here are some sample queries:
+needs to be retrieved::
+
+   SELECT *
+   FROM hudi.testdb.customer_orders FOR VERSION AS OF 8954597067493422955
+
+A different approach of retrieving historical data is to specify
+a point in time in the past, such as a day or week ago. The latest snapshot
+of the table taken before or at the specified timestamp in the query is
+internally used for providing the previous state of the table::
+
+   SELECT *
+   FROM hudi.testdb.customer_orders FOR TIMESTAMP AS OF TIMESTAMP '2022-03-23 09:59:29.803 Europe/Vienna'
```
Please use "example" as the catalog name.

ebyhr replied:
Also, could you document how users can find those versions (`8954597067493422955` and `'2022-03-23 09:59:29.803 Europe/Vienna'`) from the existing tables?

albericgenius replied:
added the description in hudi.rst. :)

ebyhr commented on docs/src/main/sphinx/connector/hudi.rst:
```diff
@@ -157,3 +157,43 @@ Here are some sample queries:
+
+    SELECT _hoodie_commit_time
+    FROM example.testdb.customer_orders
+    ORDER BY _hoodie_commit_time DESC
```
I feel recommending this query isn't a good idea.  This query causes full table scan and I suppose it means we can't time travel to where the operation was `DELETE` statement. It would be nice to provide a system table that provide commit history first in #16181

albericgenius commented:
@ebyhr i used the UI merge tool to resolve conflict issue, after that there is CI error: PR requires a rebase. Found: 1 merge.commit. and the cassandra CI test is not relative to my modification.

ebyhr commented:
Please rebase on master instead.

albericgenius commented:
> Please rebase on master instead.

Got you

nikoshet commented:
I guess this is close to getting merged 👀

albericgenius commented:
@ebyhr still need your help :)

ebyhr commented on docs/src/main/sphinx/connector/hudi.rst:
```diff
@@ -220,6 +220,46 @@ The output of the query has the following columns:
     - ``varchar``
     - Current state of the instant
 
+Rolling back to a previous version
```
@mosabua Could you please review this docs?

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:
```diff
@@ -320,6 +325,48 @@ public void testTimelineTableRedirect()
+        String after = formatter.format(new Date());
+
+        try {
+            assertThat(onTrino().executeQuery("SELECT id, name FROM hudi.default." + tableName + " FOR VERSION AS OF " + before)).hasNoRows();
```
It would be better to get the version from `$timeline` table.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:
```diff
@@ -320,6 +325,48 @@ public void testTimelineTableRedirect()
+            onHudi().executeQuery("ALTER TABLE default." + tableName + " DROP COLUMNS (new_col)");
+            onHudi().executeQuery("INSERT INTO default." + tableName + " VALUES (4, 'a4', 20, 1000)");
+            String afterDropColumn = formatter.format(new Date());
+            assertThat(onTrino().executeQuery("SELECT id, name FROM hudi.default." + tableName + " FOR VERSION AS OF " + afterDropColumn))
```
The assertions having `FOR VERSION AS OF` specifies the latest version. That's not a common scenario when using time travel. I would recommend verifying with the past version. Also, it should verify the table definition.

ebyhr commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/EnvSinglenodeHudi.java:
```diff
@@ -81,12 +82,20 @@ public EnvSinglenodeHudi(
                 "hive",
                 forHostPath(configDir.getPath("hive.properties")),
                 CONTAINER_TRINO_ETC + "/catalog/hive.properties");
+
```
nit: Revert unrelated change.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:
```diff
@@ -320,6 +325,48 @@ public void testTimelineTableRedirect()
+    @Test(groups = {HUDI, PROFILE_SPECIFIC_TESTS})
+    public void testTimeTravelQuery()
+    {
+        String tableName = "test_hudi_cow_select_session_props" + randomNameSuffix();
```
Why the table name contains `select_session_props`?

ebyhr commented on docs/src/main/sphinx/connector/hudi.rst:
```diff
@@ -220,6 +220,46 @@ The output of the query has the following columns:
+internally used for providing the previous state of the table::
+
+   SELECT *
+   FROM example.testdb.customer_orders FOR TIMESTAMP AS OF TIMESTAMP '2022-03-23 09:59:29.803 Europe/Vienna'
```
Does this PR contain a test case for `FOR TIMESTAMP AS OF TIMESTAMP`?

ebyhr reviewed (commented):
Please rebase on master to resolve confclits.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:
```diff
@@ -320,6 +325,48 @@ public void testTimelineTableRedirect()
+        formatter.setTimeZone(TimeZone.getTimeZone(ZoneOffset.UTC));
+        String before = formatter.format(new Date());
+        onHudi().executeQuery("SET hoodie.schema.on.read.enable=true");
+        createNonPartitionedTable(tableName, COW_TABLE_TYPE);
```
Please add another test for MOR table type.

ravjot28 commented:
@ebyhr  @albericgenius I have also implemented the incremental and my base is 418 version, should i raise the PR to incorporate the feature?

Please let me know if it sounds good i will raise the PR with the desired changes

tooptoop4 commented:
needs rebase

jlucking2023 commented:
any word on this getting released - we've built an EntityMatch feature into InsuranceLake which stores data in an an Entity Primary Table that uses Hudi. We'd like to be able to use Athena to do 'Time Travel'...

aviral-nayya commented:
Is this getting any love anytime soon?

mosabua commented:
:wave: @albericgenius  - this PR has become inactive. We hope you are still interested in working on it. Please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

Also fyi @codope @brandyml

mosabua commented:
Closing this PR due to inactivity. Feel free to pick it back up and continue work on it any time.
